### PR TITLE
[SNMP] Componentize SNMP traps server

### PR DESCRIPTION
**Pull request by dplepage-dd**
Split the SNMP traps server into components and register them with fx.

This is a base PR for fixing the bug behind incident 23581. These changes were already reviewed, approved, and merged as #19789, but reverted by #20773 because they caused the agent to fail when the snmp traps DB wasn't present on the filesystem even when the traps server was disabled by configuration.

This PR is a base for #20807 so that we can see what's different.

This is part of NDMII-287 and NDMII-2298.

### CODE REVIEW

File: comp/snmptraps/config/component.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+// Package config implements the configuration type for the traps server and
+// a component that provides it.
+package config
+
+// team: network-device-monitoring
+
+// Component is the component type.
+type Component interface {
+	Get() *TrapsConfig
```
Comment (hush-hush) [inline comp/snmptraps/config/component.go:14]: Could we add a doc string ?

File: comp/snmptraps/config/component.go [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023-present Datadog, Inc.
+
+// Package config implements the configuration type for the traps server and
+// a component that provides it.
+package config
+
+// team: network-device-monitoring
+
+// Component is the component type.
```
Comment (hush-hush) [inline comp/snmptraps/config/component.go:12]: Could we update this comment to better explain what this type does ?

File: comp/snmptraps/formatter/component.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+import (
+	"github.com/DataDog/datadog-agent/comp/snmptraps/packet"
+)
+
+// team: network-device-monitoring
+
+// Component is the component type.
+type Component interface {
+	FormatPacket(packet *packet.SnmpPacket) ([]byte, error)
```
Comment (hush-hush) [inline comp/snmptraps/formatter/component.go:17]: Same here for docstrings.

File: comp/snmptraps/forwarder/component.go [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+// Copyright 2023-present Datadog, Inc.
+
+// Package forwarder defines a component that receives trap data from the
+// listener component, formats it properly, and sends it to the backend.
+package forwarder
+
+// team: network-device-monitoring
+
+// Component is the component type.
+type Component interface{}
```
Comment (hush-hush) [inline comp/snmptraps/forwarder/component.go:13]: Same here for docstrings.

File: comp/snmptraps/listener/component.go [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+// parses them, and publishes messages on a channel.
+package listener
+
+import (
+	"github.com/DataDog/datadog-agent/comp/snmptraps/packet"
+)
+
+// team: network-device-monitoring
+
+// Component is the component type.
```
Comment (hush-hush) [inline comp/snmptraps/listener/component.go:16]: Same here for docstrings.

File: comp/snmptraps/oidresolver/component.go [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+// Package oidresolver resolves OIDs
+package oidresolver
+
+// team: network-device-monitoring
+
+// Component is a interface to get Trap and Variable metadata from OIDs
+type Component interface {
+	GetTrapMetadata(trapOID string) (TrapMetadata, error)
+	GetVariableMetadata(trapOID string, varOID string) (VariableMetadata, error)
```
Comment (hush-hush) [inline comp/snmptraps/oidresolver/component.go:14]: Same here for docstrings.

File: comp/snmptraps/server/component.go [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+// Package server implements a component that runs the traps server.
+// It listens for SNMP trap messages on a configured port, parses and
+// reformats them, and sends the resulting data to the backend.
+package server
+
+// team: network-device-monitoring
+
+// Component is the component type.
+type Component interface{}
```
Comment (hush-hush) [inline comp/snmptraps/server/component.go:14]: Same here for docstrings.

File: comp/snmptraps/status/component.go [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+package status
+
+// team: network-device-monitoring
+
+// Component is the component type.
+type Component interface {
+	AddTrapsPackets(int64)
+	GetTrapsPackets() int64
+	AddTrapsPacketsAuthErrors(int64)
+	GetTrapsPacketsAuthErrors() int64
```
Comment (hush-hush) [inline comp/snmptraps/status/component.go:17]: Same here for docstrings.

**Review by hush-hush — APPROVED** [review]
Looks good for files owned by agent-shared-components outside the missing docstrings on the components interfaces.

File: comp/README.md [hunk 9]
```diff
@@ -256,6 +256,47 @@ supported Datadog intakes.
 
 
 
+## [comp/snmptraps](https://pkg.go.dev/github.com/DataDog/dd-agent-comp-experiments/comp/snmptraps) (Component Bundle)
+
+*Datadog Team*: network-device-monitoring
+
+Package snmptraps implements the a server that listens for SNMP trap data
+and sends it to the backend.
```
Comment (GustavoCaso) [inline comp/README.md:264]:
```suggestion
Package snmptraps implements a server that listens for SNMP trap data
and sends it to the backend.
```

**Pull request closed without merging**